mcp/tools/tools_registry.go

```diff
@@ -0,0 +1,244 @@
+The workspace must be running with a connected agent for this to succeed.
+
+Before using this tool:
+1. Verify the workspace is running using coder_get_workspace
+2. Start the workspace if needed using coder_start_workspace
+
+Common errors:
+- Workspace not running: Start the workspace first
+- Command not allowed: Check security restrictions
+- Agent not connected: The workspace may still be starting up`),
+			mcp.WithString("workspace", mcp.Description(`The workspace ID (UUID) or name where the command will execute.
+Can be specified as either:
+- Full UUID: e.g., "8a0b9c7d-1e2f-3a4b-5c6d-7e8f9a0b1c2d" 
+- Workspace name: e.g., "dev", "python-project"
+The workspace must be running with a connected agent.
+Use coder_get_workspace first to check the workspace status.`), mcp.Required()),
+			mcp.WithString("command", mcp.Description(`The shell command to execute in the workspace.
+Commands are executed in the default shell of the workspace.
+
+Examples:
+- "ls -la" - List files with details
+- "cd /path/to/directory && command" - Execute in specific directory
+- "cat ~/.bashrc" - View a file's contents
+- "python -m pip list" - List installed Python packages
+
+Note: Very long-running commands may time out.`), mcp.Required()),
+		),
+		MakeHandler: handleCoderWorkspaceExec,
+	},
+	{
+		Tool: mcp.NewTool("coder_workspace_transition",
+			mcp.WithDescription(`Start or stop a running Coder workspace.
+If stopping, initiates the workspace stop transition.
+Only works on workspaces that are currently running or failed.
+
+If starting, initiates the workspace start transition.
+Only works on workspaces that are currently stopped or failed.
+
+Stopping or starting a workspace is an asynchronous operation - it may take several minutes to complete.
+
+After calling this tool:
+1. Use coder_report_task to inform the user that the workspace is stopping or starting
+2. Use coder_get_workspace periodically to check for completion
+
+Common errors:
+- Workspace already started/starting/stopped/stopping: No action needed
+- Cancellation failed: There may be issues with the underlying infrastructure
+- User doesn't own workspace: Permission issues`),
+			mcp.WithString("workspace", mcp.Description(`The workspace ID (UUID) or name to start or stop.
+Can be specified as either:
+- Full UUID: e.g., "8a0b9c7d-1e2f-3a4b-5c6d-7e8f9a0b1c2d"
+- Workspace name: e.g., "dev", "python-project"
+The workspace must be in a running state to be stopped, or in a stopped or failed state to be started.
+Use coder_get_workspace first to check the current workspace status.`), mcp.Required()),
+			mcp.WithString("transition", mcp.Description(`The transition to apply to the workspace.
+Can be either "start" or "stop".`)),
+		),
+		MakeHandler: handleCoderWorkspaceTransition,
+	},
+}
+
+// ToolAdder interface for adding tools to a server
+type ToolAdder interface {
+	AddTool(tool mcp.Tool, handler server.ToolHandlerFunc)
+}
```

Comment:
Thoughts on not abstracting this for now?

I'm not sure where tools would even be useful outside of MCP, so it makes me nervous to abstract this - especially since we're actually importing the `server.MCPServer` in here.

Reply:
Abstraction un-abstracted.